src/main/native/com/code_intelligence/jazzer/driver/android_tooling.h

```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright 2021 Code Intelligence GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <jni.h>
+
+#include <string>
+
+extern const char * RUNTIME_LIBRARY;
+
+extern "C" {
+    // Function to register libandroid_runtime JNI functions with java env.
```

Comment:
Instead of having this header file, you can include the header file generated by the `.hdrs` target in the `.cpp` file.

Reply:
@fmeum This is the approach I tried initially but couldn't get working.

Here is my`java_jni_library` definition, referencing the `native_lib`:

```
java_jni_library(
    name = "android_runtime",
    srcs = ["AndroidRuntime.java"],
    native_libs = ["//src/main/native/com/code_intelligence/jazzer/driver:jazzer_android_tooling"],
    visibility = [
        "//src/main/java/com/code_intelligence/jazzer:__pkg__",
        "//src/main/java/com/code_intelligence/jazzer/driver:__subpackages__",
        "//src/main/native/com/code_intelligence/jazzer/driver:__subpackages__",
    ],
    deps = [
        "//src/main/java/com/code_intelligence/jazzer/driver:opt",
        "//src/main/java/com/code_intelligence/jazzer/utils:log",
    ],
)
```


Here is `cc_jni_library` definition, referencing the `.hdrs` file in `deps` per [the rules docs](https://github.com/fmeum/rules_jni/blob/main/docs/rules.md):

```
cc_jni_library(
    name = "jazzer_android_tooling",
    srcs = ["android_tooling.cpp"],
    platforms = MULTI_PLATFORM,
    visibility = ["//src/main/java/com/code_intelligence/jazzer/android:__pkg__"],
    deps = ["//src/main/java/com/code_intelligence/jazzer/android:android_runtime.hdrs"],
)
```



The binary builds correctly but then at runtime throws the following error, indicating the headers file is not found / the JNI function can't be located. Can you please advise?


```
05-22 11:26:00.708 17993 17993 E app_process: No implementation found for void com.code_intelligence.jazzer.android.AndroidRuntime.registerNatives(byte[]) (tried Java_com_code_1intelligence_jazzer_android_AndroidRuntime_registerNatives and Java_com_code_1intelligence_jazzer_android_AndroidRuntime_registerNatives___3B) - is the library loaded, e.g. System.loadLibrary?
05-22 11:26:00.708 17993 17993 D AndroidRuntime: Shutting down VM
05-22 11:26:00.709 17993 17993 E AndroidRuntime: FATAL EXCEPTION: main
05-22 11:26:00.709 17993 17993 E AndroidRuntime: PID: 17993
05-22 11:26:00.709 17993 17993 E AndroidRuntime: java.lang.UnsatisfiedLinkError: No implementation found for void com.code_intelligence.jazzer.android.AndroidRuntime.registerNatives(byte[]) (tried Java_com_code_1intelligence_jazzer_android_AndroidRuntime_registerNatives and Java_com_code_1intelligence_jazzer_android_AndroidRuntime_registerNatives___3B) - is the library loaded, e.g. System.loadLibrary?
05-22 11:26:00.709 17993 17993 E AndroidRuntime:        at com.code_intelligence.jazzer.android.AndroidRuntime.registerNatives(Native Method)
05-22 11:26:00.709 17993 17993 E AndroidRuntime:        at com.code_intelligence.jazzer.android.AndroidRuntime.initialize(AndroidRuntime.java:37)
05-22 11:26:00.709 17993 17993 E AndroidRuntime:        at com.code_intelligence.jazzer.Jazzer.start(Jazzer.java:98)
05-22 11:26:00.709 17993 17993 E AndroidRuntime:        at com.code_intelligence.jazzer.Jazzer.main(Jazzer.java:65)
05-22 11:26:00.709 17993 17993 E AndroidRuntime:        at com.android.internal.os.RuntimeInit.nativeFinishInit(Native Method)
05-22 11:26:00.709 17993 17993 E AndroidRuntime:        at com.android.internal.os.RuntimeInit.main(RuntimeInit.java:359)
05-22 11:26:00.713 17993 17993 E AndroidRuntime: Error reporting crash
05-22 11:26:00.713 17993 17993 E AndroidRuntime: java.lang.RuntimeException: Bad file descriptor
05-22 11:26:00.713 17993 17993 E AndroidRuntime:        at android.os.BinderProxy.transactNative(Native Method)
05-22 11:26:00.713 17993 17993 E AndroidRuntime:        at android.os.BinderProxy.transact(BinderProxy.java:584)
05-22 11:26:00.713 17993 17993 E AndroidRuntime:        at android.os.IServiceManager$Stub$Proxy.checkService(IServiceManager.java:443)
05-22 11:26:00.713 17993 17993 E AndroidRuntime:        at android.os.ServiceManagerProxy.getService(ServiceManagerNative.java:63)
05-22 11:26:00.713 17993 17993 E AndroidRuntime:        at android.os.ServiceManager.rawGetService(ServiceManager.java:378)
05-22 11:26:00.713 17993 17993 E AndroidRuntime:        at android.os.ServiceManager.getService(ServiceManager.java:147)
05-22 11:26:00.713 17993 17993 E AndroidRuntime:        at android.app.ActivityManager$3.create(ActivityManager.java:5204)
05-22 11:26:00.713 17993 17993 E AndroidRuntime:        at android.app.ActivityManager$3.create(ActivityManager.java:5201)
05-22 11:26:00.713 17993 17993 E AndroidRuntime:        at android.util.Singleton.get(Singleton.java:43)
05-22 11:26:00.713 17993 17993 E AndroidRuntime:        at android.app.ActivityManager.getService(ActivityManager.java:5192)
05-22 11:26:00.713 17993 17993 E AndroidRuntime:        at com.android.internal.os.RuntimeInit$KillApplicationHandler.uncaughtException(RuntimeInit.java:160)
05-22 11:26:00.713 17993 17993 E AndroidRuntime:        at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1071)
05-22 11:26:00.713 17993 17993 E AndroidRuntime:        at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1066)
05-22 11:26:00.713 17993 17993 E AndroidRuntime:        at java.lang.Thread.dispatchUncaughtException(Thread.java:2306)
```

Reply:
The library is also loaded before the native java function is called:

```
RulesJni.loadLibrary("jazzer_android_tooling", "/com/code_intelligence/jazzer/driver");
```

Reply:
The issue was 2-fold:

1. I had a mismatched return type of `void` from the native java function and `jint` for the native cpp function. Updated `void` to `int`.
2. My native cpp function did not contain the apparently required `jclass` variable. After adding it, the function now works as expected.

Can I raise a PR to make #2 more clear, maybe by adding a comment to the `math.c` example:

https://github.com/fmeum/rules_jni/blob/7adcd41f3029d78f02eb42e5e9848d551eb1e3d7/tests/native_loader/src/main/native/com/example/math/math.c#L19

Reply:
Thanks for the responsiveness! Done.

https://github.com/fmeum/rules_jni/pull/100